src/node_report_module.cc

```diff
@@ -0,0 +1,506 @@
+#include "node_report.h"
+#include "node_internals.h"
+
+#include <sstream>
+#include <v8.h>
+
+using v8::FunctionCallbackInfo;
+using v8::HandleScope;
+using v8::Isolate;
+using v8::NewStringType;
+using v8::StackTrace;
+using v8::String;
+using v8::V8;
+using v8::Value;
+
+namespace nodereport {
+
+// Internal/static function declarations
+static void OnFatalError(const char* location, const char* message);
+bool OnUncaughtException(Isolate* isolate);
+#ifdef _WIN32
+static void PrintStackFromStackTrace(Isolate* isolate, FILE* fp);
+#else  // signal trigger functions for Unix platforms and OSX
+static void SignalDumpAsyncCallback(uv_async_t* handle);
+inline void* ReportSignalThreadMain(void* unused);
+static int StartWatchdogThread(void* (*thread_main)(void* unused));
+static void RegisterSignalHandler(int signo, void (*handler)(int),
+                                  struct sigaction* saved_sa);
+static void RestoreSignalHandler(int signo, struct sigaction* saved_sa);
+static void SignalDump(int signo);
+static void SetupSignalHandler();
+#endif
+
+// Default node-report option settings
+static unsigned int nodereport_events = NR_APICALL;
+static unsigned int nodereport_verbose = 0;
+#ifdef _WIN32  // signal trigger not supported on Windows
+static unsigned int nodereport_signal = 0;
+#else  // trigger signal supported on Unix platforms and OSX
+static unsigned int nodereport_signal = SIGUSR2;  // default signal is SIGUSR2
+static int report_signal = 0;  // atomic for signal handling in progress
+static uv_sem_t report_semaphore;  // semaphore for hand-off to watchdog
+static uv_async_t nodereport_trigger_async;  // async handle for event loop
+static uv_mutex_t node_isolate_mutex;  // mutex for watchdog thread
+static struct sigaction saved_sa;  // saved signal action
+#endif
+
+// State variables for v8 hooks and signal initialisation
+static bool exception_hook_initialised = false;
+static bool error_hook_initialised = false;
+static bool signal_thread_initialised = false;
+
+static Isolate* node_isolate;
```

Comment:
Uh … what happens if there is more than one isolate?

Reply:
Removed the global isolate variable and confined to methods, but brings up a topic of discussion: How do we want to deal with the report capability for workers? There are scenarios (for example when an external signal is received for report generation, which Isolate / worker should respond to that)

As of I have made sure thread-saftey and confined the signal registration to the main Isolate in `src/node.cc` , full multi-isolate support could be achieved through a follow- up PR, if you agree. Please let me know.